runtime/lua/vim/treesitter/query.lua

```diff
@@ -208,16 +208,54 @@ function M.set_query(...)
   M.set(...)
 end
 
+---@class SetQueryOpts
+---@field inherits (string|string[]|nil)
+---@field extends (boolean|nil)
+
 --- Sets the runtime query named {query_name} for {lang}
 ---
---- This allows users to override any runtime files and/or configuration
+--- This allows users to override or extend any runtime files and/or configuration
 --- set by plugins.
 ---
 ---@param lang string Language to use for the query
 ---@param query_name string Name of the query (e.g., "highlights")
----@param text string Query text (unparsed).
-function M.set(lang, query_name, text)
-  explicit_queries[lang][query_name] = M.parse(lang, text)
+---@param text string Query text (unparsed)
+---@param opts? SetQueryOpts (table) Optional keyword arguments
+---   - inherits (string|string[]|nil) Language(s) this query should inherit
+---     (see |treesitter-query-modeline-inherits|)
+---   - extends (boolean|nil) If `true`, the query should be used as an extension query
+---     (see |treesitter-query-modeline-extends|)
+function M.set(lang, query_name, text, opts)
+  local query ---@type vim.treesitter.Query
+
+  if not opts then
+    query = M.parse(lang, text)
+  else
+    local query_files = {}
+
+    if opts.inherits then
+      local inherits = opts.inherits
+      if type(inherits) == 'string' then
+        inherits = { inherits }
+      end
+      ---@cast inherits string[]
+
+      for _, inherited in ipairs(inherits) do
+        local files = M.get_files(inherited, query_name, true)
+        vim.list_extend(query_files, files)
+      end
+    end
+
+    if opts.extends then
+      local files = M.get_files(lang, query_name)
+      vim.list_extend(query_files, files)
+    end
```

Comment:
Just to clarify, you're referring to multiple extensions being applied by calling `set` multiple times, correct? So multiple plugins can extend the same query? If so, reading runtime files and parsing can be delayed until `vim.treesitter.query.get` is called instead of within `set`.

I noticed one other thing while adding the test for `get` after `set`. `vim.treesitter.query.get` is now memoized, so a call to `get` isn't guaranteed to include the most recent `set` if `get` was previously called. Should the memoize cache be invalidated for a given query after a call to `set`?